### ISSUE TITLE
Issue in OffRouteDetector

### ISSUE THREAD

**Issue by richjhart** [mapbox/mapbox-navigation-android#1081]
**Android API:** N/A
**Mapbox Navigation SDK version:** 0.15.0

### Steps to trigger behavior

 1. Start navigation
 2. Move slowly along the route (walking pace can be enough), such that the distance to maneuver stays basically the same
3. Wait for 4 GPS readings

### Expected behavior

- Navigation continues normally

### Actual behavior

- Rerouting is triggered

I believe this is caused by an error in OffRouteDetector.movingAwayFromManeuver(). it checks that:
distancesAwayFromManeuver.peekLast()
      - distancesAwayFromManeuver.peekFirst() < MINIMUM_BACKUP_DISTANCE_FOR_OFF_ROUTE
I believe there is an error here.
- The queue will already be cleared (later in the code) if the distance is decreasing.
- Therefore the queue from oldest to newest must be increasing (or it could remain constant)
- peekLast() takes the oldest value, peekFirst() takes the newest, so peekLast() - peekFirst() will always be <= 0
- Which means this check will always be true

I assume it is meant to check if the user has gone "more than 50m backwards" before triggering, in which case the check should be:
peekFirst() - peekLast() > MINIMUM_BACKUP_DISTANCE_FOR_OFF_ROUTE

Note that because distancesAwayFromManeuver are integers, even slightly decreasing distances will still trigger this - as long as they round to the same integer.

e.g. distances of 210.40, 210.18, 210.34 and 209.94 will trigger this.

As will distances of 200, 210, 220, 230 (which is backwards, but as per the above description, I assume is meant to be filtered).

Also note the 4th distance seems to be pointless. It will always trigger then anyway because the "return true" happens before there is a chance to clear if it's back in the right direction. If that's intended, it's not clear why it doesn't just trigger on the previous call.

**Comment by danesfeder** [comment]
Hey @richjhart 👋 thanks for the great write-up here.  We will take a look at this logic in-depth and get back to you here on this ticket.  At first glance, all of your points look completely valid to me.

**Comment by dsilvera** [comment]
I have the same problem.
Thank you

**Comment by mclucky** [comment]
We've the same problem.

**Comment by siavashabdoli** [comment]
I have same problem on v16. here is logcat:
`8-17 17:14:31.778 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:33.744 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:34.827 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:36.764 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: true
08-17 17:14:37.815 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:38.874 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:40.853 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:42.788 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:44.785 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: true
08-17 17:14:45.800 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:46.820 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:48.788 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:49.848 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: true
08-17 17:14:52.904 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:54.819 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: isUserOffRoute: false
08-17 17:14:55.823 1978-2089/com.mapbox.services.android.navigation.testapp D/RouteProcessorHandlerCallback: `

after first offRoute it should print true for all logs but as you can see sometimes it detect user is not offroute and Rerouting is not triggered.

**Comment by siavashabdoli** [comment]
**UPDATE**
I think I found the reason and some added information for debuging:
the problem would occur in last leg. the `checkForArrivalEvent` method in `NavigationEventDispatcher` consider it as Arrival and delete the listener so when offroute occur there is no listener to invoke reroute request

**Comment by danesfeder** [comment]
@richjhart @siavashabdoli do ya'll mind retesting with `0.20.0`?

**Comment by richjhart** [comment]
@danesfeder If I get time I'll have a look. But we've actually taken the opportunity to build our own off route engine, because we wanted extra functionality. I'll have to see how easy it is to go back to the original.

**Comment by danesfeder** [comment]
@richjhart okay great, thank you.  @siavashabdoli if you don't mind testing as well, I'd just like to confirm this is no longer and issue with the new logic.

**Comment by Guardiola31337** [comment]
Hey @siavashabdoli 👋 were you able to retest? It'd be 💯 if you could report back confirming that the issue is gone. Thanks a lot!

I also want to take the opportunity to let you know that we released version [`0.22.0`](https://github.com/mapbox/mapbox-navigation-android/releases/tag/v0.22.0) so I encourage you to upgrade because there are a bunch of improvements and new features plus lots of bug fixes since `0.16.0`.

**Comment by siavashabdoli** [comment]
Hi sorry I didn't see the thread. I will test in tomorrow and let you know 🙏🏻

**Comment by Guardiola31337** [comment]
Sweet! Thanks for the update @siavashabdoli 🙇

**Comment by danesfeder** [comment]
Hey @siavashabdoli @richjhart I'm going to go ahead and close this as resolved.  @richjhart once we unblock you from upgrading and if you re-test and still see issues, please feel free to cut a new ticket and we will continue to dig on this.  Thanks!

**Comment by danesfeder** [comment]
@richjhart this should be right behind our current push for offline routing - I can't give a specific timeline but this is definitely high priority.  We will be sure to update your ticket regarding this once we start moving.  Thanks as always for your patience 👍
